src/java.base/share/classes/jdk/internal/foreign/abi/VMStorage.java

```diff
@@ -26,32 +26,22 @@
 
 import java.util.Objects;
 
-public class VMStorage {
-    /**
-     * Type of storage. e.g. stack, or which register type (GP, FP, vector)
-     */
-    private final byte type;
-
-    /**
-     * The (on stack) size in bytes when type = stack, a register mask otherwise.
-     * The register mask indicates which segments of a register are used.
-     */
-    private final short segmentMaskOrSize;
-
-    /**
-     * The index is either a register number within a type, or
-     * a stack offset in bytes if type = stack.
-     * (a particular platform might add a bias to this in generate code)
-     */
-    private final int indexOrOffset;
-
-    private final String debugName;
+/**
+ * @param type              the type of storage. e.g. stack, or which register type (GP, FP, vector)
+ * @param segmentMaskOrSize The (on stack) size in bytes when type = stack, a register mask, otherwise
+ *                          the register mask indicates which segments of a register are used.
+ * @param indexOrOffset     The index is either a register number within a type, or
+ *                          a stack offset in bytes if type = stack.
+ *                          (a particular platform might add a bias to this in generate code)
+ * @param debugName         the debug name to use
+ */
+public record VMStorage(byte type,
```

Comment:
I think this should stay a `class`. VMStorage should not be constructed directly, but one of the factory methods should be used. But, the change to `record` forces the constructor to be public.

Reply:
Doesn't the same reasoning hold for Binding? E.g. there's a builder there, no?

Reply:
For Binding there's a builder, but it just calls the constructor of the respective Binding class. i.e. there's no additional instantiation logic encapsulated that can lead to a malformed object. Although, it is possible to lead to malformed recipes.

VMStorage does encapsulate some instantiation logic, namely the debug name for stack storage. Though, to be fair, there used to be more of that in an earlier version where I also checked that the wrong `type` was not being used. So, maybe it's time to revisit 🤔 

Thinking about it now, I think the real encapsulation happens in the architecture classes (e.g. https://github.com/openjdk/panama-foreign/blob/cb272592d5043aaa2d9f013eac37620de50418ef/src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/AArch64Architecture.java#L139) where the type of stack storage is actually known.

Maybe it would be better to make `VMStorage` just a dumb record, and drop the factory methods altogether.

Reply:
although this seems to deserve its own targeted PR

Reply:
Yes. Let's do this in a separate PR.